Add preference for membership notifications in UserEmailPreferencesModel and relevant changes

prasanna08 opened the pull request:
First milestone of #1357 
Changes:
  - Modifies UserEmailPreferencesModel
  - Modifies user_services.update_email_preferences()
  - Modifies user_services.get_email_preferences()
  - Add Default value for membership notifications in feconf
  - A backend test is also added in user_services_test.py

seanlip commented:
@prasanna08, the backend tests seem to be failing on Travis. Did you run them locally?

prasanna08 commented:
Ohh, profile_test are failing since they check return value of user_services.get_email_preferences() method and I have modified its return value.
I have fixed it locally

seanlip commented on core/domain/user_services.py:
```diff
@@ -406,7 +406,10 @@ def record_user_started_state_editor_tutorial(user_id):
 
 
-def update_email_preferences(user_id, can_receive_email_updates):
+def update_email_preferences(
```
Just to clarify -- say a user has set their membership preference to False. Then they change "can_receive_email_updates" through their Preferences page. Wouldn't this reset their membership preferences to True, given how this function is implemented?

Instead, perhaps there ought to be two separate functions, one for each type of update.

prasanna08 replied:
You see when new user is added this function is called with only first 2 arguments. Since at time of sign up we are not going to give user a choice for receiving this type of emails, I thought that 3rd argument should ensure that receiving should be set to default value. Thus adding 3rd argument as default arg maintains older implementation of this function. If it is explicit then we need to pass some value to it which will be default value. So rather than modifying function call I thought why not change it so it will work for older calls as well as for new features both.

Now what you are saying in this comment. No value won't change like that. The calls made for this function will be changed during the milestone 2. Generally what I was thinking that, whichever property user wants to change, we will simply transfer values for both properties from front end to backend. After that we will call this function using 3 arguments. So you see changes will be reflected in Model. But that will be done in milestone 2 because if we do it now in this commit then we are yet to implement frontend facility for user to be able to change value but we are changing backend facility and it will result in errors.

Now the other point you said that we use a different function for this. But then we will have 2 different functions to change same model's different properties and will not it lead to some confusion ?
I think we should go implement it the way I did because it does not break any current workings of oppia and in milestone 2 we are going to modify function call from 2 args to 3 args.

So how does that sound ?

seanlip replied:
I think I see the reasoning. The general underlying point here is that this function is called only from signup or the preferences page. In all cases, though, it's always going to be called with all the arguments, so the error scenario I was worried about above will not occur. Is that right?

In that case, I think what you're saying makes sense, and we can just have one function for the email preferences update. However, I would still like to not have a default third arg -- instead, it would be better to modify the callsites to pass in the default feconf arg. This is because I want this function to actually raise an error when it's called with fewer args than expected, and not do a silent substitution. Changes for milestone 2 will then touch only the PreferencesHandler controller, and not this layer. Would that be OK?

prasanna08 replied:
Yes thats right. This function is called only in those cases. Now we can make third arg explicit. We will pass defaulr feconf arg in signup. But what about preferences handler when it calls this function. I can think of two way :

1. We nake 3rd arg explicit and pass default feconf as arg wherever function is called. We will need a placeholder 3rd arg in preference handler which will be changed with the value received from frontend in 2nd milestone.

2. We keep it default and change 3rd arg in 2nd milestone because 2nd milestone will change this function calls. In former case we will have to change function args right now in this commit so that oppia stays in stable state

So which way do you prefer ?

seanlip replied:
Yep, that's a good analysis. I prefer the first way.

seanlip commented on core/domain/user_services.py:
```diff
@@ -406,7 +406,10 @@ def record_user_started_state_editor_tutorial(user_id):
+def update_email_preferences(
+        user_id, can_receive_email_updates,
+        can_receive_membership_email=(
+            feconf.DEFAULT_MEMBERSHIP_EMAIL_PREFERENCE)):
```
I think it's better to pass in this arg explicitly.

seanlip commented on core/domain/user_services_test.py:
```diff
@@ -132,6 +132,40 @@ def test_get_user_id_from_username(self):
             user_services.get_user_id_from_username('fakeUsername'))
 
+    def test_set_and_get_user_email_preferences(self):
+        user_id = 'someUser'
```
'userId'

prasanna08 replied:
It is `user_id` in all previous test units. Why different for this one ?

seanlip replied:
I meant 'someUser' --> 'userId'.

Having said that, I realize there are other tests in this file which already use 'someUser', so I retract my comment. Thanks for pushing back!

seanlip commented on core/domain/user_services_test.py:
```diff
@@ -132,6 +132,40 @@ def test_get_user_id_from_username(self):
+        user_services.get_or_create_user(user_id, user_email)
+        user_services.set_username(user_id, username)
+
+        # When UserEmailPreferencesModel is yet to be created value of
```
When UserEmailPreferencesModel is yet to be created, the value returned by get_email_preferences() should be True.

seanlip commented on core/domain/user_services_test.py:
```diff
@@ -132,6 +132,40 @@ def test_get_user_id_from_username(self):
+        user_services.update_email_preferences(
+            user_id, feconf.DEFAULT_EMAIL_UPDATES_PREFERENCE)
+
+        # When nothing is specified for membership email notifications
```
The user sets their membership email preference to True.

seanlip replied:
(and, use a different function)

seanlip commented on core/domain/user_services_test.py:
```diff
@@ -132,6 +132,40 @@ def test_get_user_id_from_username(self):
+            email_preferences['can_receive_membership_email'],
+            feconf.DEFAULT_MEMBERSHIP_EMAIL_PREFERENCE)
+
+        # Change preference for receiving membership emails to False
```
The user sets their membership email preference to False.

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -96,6 +96,11 @@ class UserEmailPreferencesModel(base_models.BaseModel):
 
+    # The user's preference for receiving email when user is added as a member
+    # in exploration. This is set to True when user has never set a preference.
+    membership_notifications = ndb.BooleanProperty(
```
I've been thinking about this, and I think a better name would be editor_role_notifications. I don't think we currently have the concept of "member", or that we want to introduce the new terminology yet; I found 'role' in core/domain/rights_manager.py. What do you think?

prasanna08 replied:
Hmm well if it is like that then no problem. Since whatever we name it, it must be able to clarify why it is there. I do not have problem with it. Will do the change in next commit.

Thanx

prasanna08 commented:
HI @seanlip I have addressed all your suggested changes in this new commit.

seanlip commented on core/domain/user_services_test.py:
```diff
@@ -132,6 +132,41 @@ def test_get_user_id_from_username(self):
+        user_services.set_username(user_id, username)
+
+        # When UserEmailPreferencesModel is yet to be created,
+        # the value returned by get_email_preferences() should be True.
```
... should be the default value.

seanlip replied:
Actually -- on second thoughts, the comment here can just be "The UserEmailPreferencesModel is not yet created." It's quite easy to infer what's going on, given that context.

seanlip commented on core/domain/user_services_test.py:
```diff
@@ -132,6 +132,41 @@ def test_get_user_id_from_username(self):
+            email_preferences['can_receive_editor_role_email'],
+            feconf.DEFAULT_EDITOR_ROLE_EMAIL_PREFERENCE)
+
+        # When UserEmailPreferencesModel is created for user,
```
Similarly: "The user retrieves their email preferences, thus initializing a UserEmailPreferencesModel instance."

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -96,6 +96,11 @@ class UserEmailPreferencesModel(base_models.BaseModel):
+    # The user's preference for receiving email when user is added as a member
+    # in exploration. This is set to True when user has never set a preference.
+    editor_role_notifications = ndb.BooleanProperty(
+        indexed=False, default=feconf.DEFAULT_EDITOR_ROLE_EMAIL_PREFERENCE)
```
I suggest indexed=True -- we may need to query against this in the future.

seanlip commented on feconf.py:
```diff
@@ -155,6 +155,8 @@ def get_empty_ratings():
 CAN_SEND_EMAILS_TO_USERS = False
 # Whether to send email updates to a user who has not specified a preference.
 DEFAULT_EMAIL_UPDATES_PREFERENCE = False
+# Whther to send membership email when user who has not specified a preference.
```
Whether to send an invitation email when the user is granted new role permissions in an exploration.

seanlip commented:
Just some documentation tweaks, but everything else looks good, and this can be merged after that. Thanks @prasanna08!

seanlip commented:
(Also, sorry, please refresh -- I removed one of the comments, on a second reading.)

prasanna08 commented:
Should I merge the develop in this branch ?

seanlip commented:
Yes, I think so -- Github is showing me a merge error.

seanlip commented on core/domain/user_services_test.py:
```diff
@@ -132,6 +132,41 @@ def test_get_user_id_from_username(self):
+            email_preferences['can_receive_editor_role_email'],
+            feconf.DEFAULT_EDITOR_ROLE_EMAIL_PREFERENCE)
+
+        # The user retrieves their email preferences. This initializes,
```
nit: remove comma at end of line.

prasanna08 replied:
Oh Sorry. But the thing is I did not see this comment before but now it is there.
Something is wrong with either me or github.

seanlip commented on core/domain/user_services_test.py:
```diff
@@ -132,6 +132,41 @@ def test_get_user_id_from_username(self):
+            feconf.DEFAULT_EDITOR_ROLE_EMAIL_PREFERENCE)
+
+        # The user retrieves their email preferences. This initializes,
+        # a UserEmailPreferencesModel instance with the default values
```
nit: add full stop to end of sentence.

prasanna08 replied:
(Ah, my bad) As it seems I am not good at giving nicer comments. :smile:
Its LGTM so should I add commit fixing above or not ?

prasanna08 commented:
@seanlip  All good and done. Merged develop and also fixed that full stop in comment.
Now you should be able to merge it.

Thanx

seanlip commented:
No worries :) It's an attention-to-detail thing. Speaking of which, though, I don't think you removed the trailing comma mentioned [here](https://github.com/oppia/oppia/pull/1702#discussion_r59125613) -- could you fix it? Also, next time, please check that you've addressed all comments before committing.

Otherwise, looks good! I'll merge when the Travis checks pass.

seanlip commented:
> Oh Sorry. But the thing is I did not see this comment before but now it is there. Something is wrong with either me or github. 

Hm, weird. You didn't get an email? Generally, GitHub has notification settings that you can adjust; I receive emails for all comments made on the repo, but I think you can subscribe to things like individual threads.

In any case -- the changes look good! I'll merge now. Thanks @prasanna08!
